config/init_server_creds.go

```diff
@@ -503,44 +567,186 @@ func GenerateCert() error {
+	}
+
+	// Ensure input directory dir exists, if not, create it with proper permissions
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := createDirForKeys(dir); err != nil {
+			return nil, errors.Wrapf(err, "failed to create directory and set permissions: %s", dir)
+		}
+	}
+	// Traverse the directory for .pem files in lexical order
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !d.IsDir() && filepath.Ext(d.Name()) == ".pem" {
+			// Parse the private key in this file and add to the in-memory keys map
+			key, err := loadSinglePEM(path)
+			if err != nil {
+				log.Warnf("Failed to load key %s: %v", path, err)
+				return nil // Skip this file and continue
+			}
+
+			latestKeys[key.KeyID()] = key
+
+			// Update the most recent key based on lexicographical order of filenames
+			if mostRecentFileName == "" || d.Name() > mostRecentFileName {
+				mostRecentFileName = d.Name()
+				mostRecentKey = key
+			}
+		}
+		return nil
+	})
+
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to add alg specification to key header")
+		return nil, errors.Wrapf(err, "failed to traverse directory %s that stores private keys", dir)
+	}
+
+	// Create a new private key and set as issuer key when neither legacy private key at IssuerKey
+	// nor any .pem file at IssuerKeysDirectory exists
+	if len(latestKeys) == 0 || mostRecentKey == nil {
+		newKey, err := generatePEMandSetIssuerKey(dir)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create a new .pem file to save private key")
+		}
+		return newKey, nil
+	}
+
+	// Save current key and all up-to-date valid private keys and the in-memory issuerKeys
+	newKeys := IssuerKeys{
+		CurrentKey: mostRecentKey,
+		AllKeys:    latestKeys,
 	}
+	issuerKeys.Store(&newKeys)
+	log.Debugf("Set private key %s as the issuer key", mostRecentKey.KeyID())
+
+	return mostRecentKey, nil
+}
 
-	// Assign key id to the private key so that the public key obtainer thereafter
-	// has the same kid
-	err = jwk.AssignKeyID(key)
+// Create a new .pem file (combining GeneratePrivateKey and LoadPrivateKey functions)
+func GeneratePEM(dir string) (jwk.Key, error) {
+	// Generate a unique filename using a POSIX mkstemp-like logic
+	// Create a temp file, store its filename, then immediately delete this temp file
+	filenamePattern := fmt.Sprintf("pelican_generated_%d_*.pem",
+		time.Now().UnixNano())
+	if err := createDirForKeys(dir); err != nil {
+		return nil, errors.Wrapf(err, "failed to create directory and set permissions: %s", dir)
+	}
+	tempFile, err := os.CreateTemp(dir, filenamePattern)
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to assign key ID to private key")
+		return nil, errors.Wrap(err, "failed to remove temp file")
+	}
+	keyPath := tempFile.Name()
+	if err := tempFile.Close(); err != nil {
+		return nil, errors.Wrap(err, "failed to close temp file")
+	}
+	if err := os.Remove(keyPath); err != nil {
```

Comment:
This introduces a race condition -- another file can be created after the removal but before it's rewritten.

Consider having a helper function take an open file handle (still making sure the mode is correct) and generate the key and refactor the existing `GeneratePrivateKey` to use that.